{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "30372e34-905f-46fd-9e93-5a50343048f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import sys\n",
    "from optparse import OptionParser\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.collections import LineCollection\n",
    "%matplotlib inline\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.cm as cm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3aa51d6e-70f6-42d1-bec3-396ccc2f3b70",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GC_content_window(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC content of input sequence\n",
    "    \"\"\"\n",
    "\n",
    "    gc = sum(s.count(x) for x in ['G','C','g','c','S','s'])\n",
    "    sl = sum(s.count(x) for x in ['A','a','T','t','G','C','g','c','S','s'])\n",
    "    try:\n",
    "        gc_content = gc/float(sl)\n",
    "    except ZeroDivisionError:\n",
    "        gc_content = 0\n",
    "    return round(gc_content,4) \n",
    "\n",
    "def GC_skew_window(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC skew of a sequence\n",
    "    \"\"\"\n",
    "\n",
    "    g = s.count('G')+s.count('g')\n",
    "    c = s.count('C')+s.count('c')\n",
    "\n",
    "    try:\n",
    "        skew = (g-c)/float(g+c)\n",
    "    except ZeroDivisionError:\n",
    "        skew = 0\n",
    "    return round(skew,4)\n",
    "\n",
    "def AT_skew_window(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC skew of a sequence\n",
    "    \"\"\"\n",
    "\n",
    "    a = s.count('A')+s.count('a')\n",
    "    t = s.count('T')+s.count('t')\n",
    "\n",
    "    try:\n",
    "        skew = (a-t)/float(a+t)\n",
    "    except ZeroDivisionError:\n",
    "        skew = 0\n",
    "    return round(skew,4)\n",
    "\n",
    "def AG_content_window(s):\n",
    "\n",
    "    \"\"\"\n",
    "    Return AG content of a sequence\n",
    "    \"\"\"\n",
    "    \n",
    "    ag = sum(s.count(x) for x in ['A','G','a','g', 'R', 'r'])\n",
    "    sl = sum(s.count(x) for x in ['A','a','T','t','G','C','g','c','S','s'])\n",
    "    try:\n",
    "        ag_content = ag/float(sl)\n",
    "    except ZeroDivisionError:\n",
    "        ag_content = 0\n",
    "    return round(ag_content,4)\n",
    "\n",
    "def GC_content_window_NR(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC content of input sequence\n",
    "    \"\"\"\n",
    "\n",
    "    gc = sum(s.count(x) for x in ['G','C','S'])\n",
    "    sl = sum(s.count(x) for x in ['A','T','G','C','S'])\n",
    "    try:\n",
    "        gc_content = gc/float(sl)\n",
    "    except ZeroDivisionError:\n",
    "        gc_content = 0\n",
    "    return round(gc_content,4) \n",
    "\n",
    "def GC_skew_window_NR(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC skew of a sequence\n",
    "    \"\"\"\n",
    "\n",
    "    g = s.count('G')\n",
    "    c = s.count('C')\n",
    "\n",
    "    try:\n",
    "        skew = (g-c)/float(g+c)\n",
    "    except ZeroDivisionError:\n",
    "        skew = 0\n",
    "    return round(skew,4)\n",
    "\n",
    "def AT_skew_window_NR(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC skew of a sequence\n",
    "    \"\"\"\n",
    "\n",
    "    a = s.count('A')\n",
    "    t = s.count('T')\n",
    "\n",
    "    try:\n",
    "        skew = (a-t)/float(a+t)\n",
    "    except ZeroDivisionError:\n",
    "        skew = 0\n",
    "    return round(skew,4)\n",
    "\n",
    "def AG_content_window_NR(s):\n",
    "\n",
    "    \"\"\"\n",
    "    Return AG content of a sequence\n",
    "    \"\"\"\n",
    "    \n",
    "    ag = sum(s.count(x) for x in ['A','G','R'])\n",
    "    sl = sum(s.count(x) for x in ['A','T','G','C','S'])\n",
    "    try:\n",
    "        ag_content = ag/float(sl)\n",
    "    except ZeroDivisionError:\n",
    "        ag_content = 0\n",
    "    return round(ag_content,4)\n",
    "\n",
    "def GC_content_window_R(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC content of input sequence\n",
    "    \"\"\"\n",
    "\n",
    "    gc = sum(s.count(x) for x in ['g','c','s'])\n",
    "    sl = sum(s.count(x) for x in ['a','t','g','c','s'])\n",
    "    try:\n",
    "        gc_content = gc/float(sl)\n",
    "    except ZeroDivisionError:\n",
    "        gc_content = 0\n",
    "    return round(gc_content,4) \n",
    "\n",
    "def GC_skew_window_R(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC skew of a sequence\n",
    "    \"\"\"\n",
    "\n",
    "    g = s.count('g')\n",
    "    c = s.count('c')\n",
    "\n",
    "    try:\n",
    "        skew = (g-c)/float(g+c)\n",
    "    except ZeroDivisionError:\n",
    "        skew = 0\n",
    "    return round(skew,4)\n",
    "\n",
    "def AT_skew_window_R(s):\n",
    "    \n",
    "    \"\"\"\n",
    "    Return GC skew of a sequence\n",
    "    \"\"\"\n",
    "\n",
    "    a = s.count('a')\n",
    "    t = s.count('t')\n",
    "\n",
    "    try:\n",
    "        skew = (a-t)/float(a+t)\n",
    "    except ZeroDivisionError:\n",
    "        skew = 0\n",
    "    return round(skew,4)\n",
    "\n",
    "def AG_content_window_R(s):\n",
    "\n",
    "    \"\"\"\n",
    "    Return AG content of a sequence\n",
    "    \"\"\"\n",
    "    \n",
    "    ag = sum(s.count(x) for x in ['a','g', 'r'])\n",
    "    sl = sum(s.count(x) for x in ['a','t','g','c','s'])\n",
    "    try:\n",
    "        ag_content = ag/float(sl)\n",
    "    except ZeroDivisionError:\n",
    "        ag_content = 0\n",
    "    return round(ag_content,4)\n",
    "\n",
    "def null_value_window(s):\n",
    "    \"\"\"\n",
    "    return number of Ns in a sequence\n",
    "    \"\"\"\n",
    "    n = sum(s.count(x) for x in ['N'])\n",
    "    n_percent = n/float(len(s))\n",
    "    return round(n_percent, 4)\n",
    "\n",
    "def hex_to_rgba(hex_string):\n",
    "    rgb = mcolors.hex2color(hex_string)\n",
    "    return rgb+(1,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1f36e9ac-2045-490c-bed6-0de80d6216ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nnumChromosomes = 19\\nfilename = 'input_data/genomes/Bfl.fna'\\ngtffile = 'input_data/gene_rows/Bfl.gtf'\\nSCfile = 'input_data/gene_rows_SC/Bfl_coordinates.tsv'\\n\""
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "window = 20000\n",
    "step = 20000\n",
    "nThreshold = 0.5\n",
    "graphShape = (10,12)\n",
    "\n",
    "color_mapping = {\"A\": \"#A6CDE3\", \n",
    "                  \"B\": \"#089E78\", \n",
    "                  \"C\": \"#F19799\", \n",
    "                  \"D\": \"#F7F09B\", \n",
    "                  \"E\": \"#AF5B2A\",\n",
    "                  \"F\": \"#DA6227\", \n",
    "                  \"G\": \"#7671B0\", \n",
    "                  \"H\": \"#E42C8A\", \n",
    "                  \"I\": \"#65A744\", \n",
    "                  \"J\": \"#C9B2D3\",\n",
    "                  \"K\": \"#E6AB22\", \n",
    "                  \"L\": \"#A4782B\", \n",
    "                  \"M\": \"#90CDC3\", \n",
    "                  \"N\": \"#686868\", \n",
    "                  \"O\": \"#EF7E25\",\n",
    "                  \"P\": \"#BDB9DA\", \n",
    "                  \"Q\": \"#FED832\", \n",
    "                  \"R\": \"#00008B\",\n",
    "                  \"S\": \"#FED832\"}\n",
    "\"\"\"\n",
    "numChromosomes = 19\n",
    "filename = 'input_data/genomes/Bfl.fna'\n",
    "gtffile = 'input_data/gene_rows/Bfl.gtf'\n",
    "SCfile = 'input_data/gene_rows_SC/Bfl_coordinates.tsv'\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a551e864-7536-4d83-aec3-5ffc915917d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nrecords = SeqIO.parse(filename, \\'fasta\\')\\ndata = [(record.id, str(record.seq), len(record.seq)) for record in records]\\nseqobj = pd.DataFrame(data, columns=[\\'id\\', \\'seq\\', \\'seqlen\\'])\\nseqobj = seqobj.sort_values(\\'seqlen\\', ascending=False).head(numChromosomes)\\n\\ncolumns = [\"sequence\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\\ngtf_dataframe = pd.read_csv(gtffile, sep = \\'\\t\\', comment = \\'#\\', header = None, names = columns, dtype={\\'start\\': int, \\'end\\': int})\\n\\nSCcolumns = [\"index\", \"Junk\", \"sequence\", \"start\", \"end\", \"color_key\"]\\nSC_dataframe = pd.read_csv(SCfile, sep = \\'\\t\\', comment = \\'#\\', header = None, names = SCcolumns, dtype={\\'start\\': int, \\'end\\': int})\\nSC_dataframe = SC_dataframe[SC_dataframe[\\'color_key\\'].isin([\"A\",\"B\",\"C\",\"D\",\"E\",\"F\",\"G\",\"H\",\"I\",\"J\",\"K\",\"L\",\"M\",\"N\",\"O\",\"P\",\"Q\",\"R\",\"S\"])]\\nSC_dataframe[\\'color\\'] = SC_dataframe[\\'color_key\\'].map(color_mapping)\\n'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "records = SeqIO.parse(filename, 'fasta')\n",
    "data = [(record.id, str(record.seq), len(record.seq)) for record in records]\n",
    "seqobj = pd.DataFrame(data, columns=['id', 'seq', 'seqlen'])\n",
    "seqobj = seqobj.sort_values('seqlen', ascending=False).head(numChromosomes)\n",
    "\n",
    "columns = [\"sequence\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "gtf_dataframe = pd.read_csv(gtffile, sep = '\\t', comment = '#', header = None, names = columns, dtype={'start': int, 'end': int})\n",
    "\n",
    "SCcolumns = [\"index\", \"Junk\", \"sequence\", \"start\", \"end\", \"color_key\"]\n",
    "SC_dataframe = pd.read_csv(SCfile, sep = '\\t', comment = '#', header = None, names = SCcolumns, dtype={'start': int, 'end': int})\n",
    "SC_dataframe = SC_dataframe[SC_dataframe['color_key'].isin([\"A\",\"B\",\"C\",\"D\",\"E\",\"F\",\"G\",\"H\",\"I\",\"J\",\"K\",\"L\",\"M\",\"N\",\"O\",\"P\",\"Q\",\"R\",\"S\"])]\n",
    "SC_dataframe['color'] = SC_dataframe['color_key'].map(color_mapping)\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4fcc5719-aaf5-4865-b1de-8aaf9180d9a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n#Add in remove repeats\\nfig, ax = plt.subplots(figsize=(8,8))\\ncolors = cm.viridis(np.linspace(0,1, numChromosomes))\\ngenome = \"\"\\ngcR = []\\ngc = []\\nj=0\\nfor index, row in seqobj.iterrows(): \\n        pos_array = []\\n        gc_content_value_array = []\\n        gc_content_value_array_NR = []\\n        gc_content_value_array_R = []\\n    \\n        n_percent_array = []\\n        name = row[\\'id\\']\\n        seq = row[\\'seq\\']\\n        for i in range(0,len(seq),step):\\n            subseq = seq[i:i+window]\\n            gc_content = (GC_content_window(subseq))\\n            gc_content_NR = (GC_content_window_NR(subseq))\\n            gc_content_R = (GC_content_window_R(subseq))\\n            n_percent = (null_value_window(subseq))\\n            #start = (i + 1 if (i+1<=len(seq)) else i)\\n            #end = ( i + step if (i+ step<=len(seq)) else len(seq))\\n            if n_percent < nThreshold:\\n                gc_content_value_array.append(gc_content)\\n                gc_content_value_array_NR.append(gc_content_NR)\\n                gc_content_value_array_R.append(gc_content_R)\\n        avg = np.average(gc_content_value_array_R)\\n        ax.scatter(gc_content_value_array_NR, gc_content_value_array_R,color=colors[j], label=str(name), alpha = 0.7)\\n        #ax.plot([0,1],[avg,avg], color=colors[index], linestyle=\"dashed\")\\n        genome = genome+seq\\n        j+=1\\nax.plot([0,1],[0,1],color=\\'black\\',linestyle=\"dashed\")\\navg = GC_content_window_NR(genome)\\navgR = GC_content_window_R(genome)\\nax.plot([0,1],[avgR,avgR], color=\\'silver\\', linestyle=\"dashed\", label=\"Average GC Content of Repeats: \"+str(np.round(avgR,3)))\\nax.plot([avg,avg],[0,1], color=\\'gold\\', linestyle=\"dashed\", label=\"Average GC Content of Non Repeats: \"+str(np.round(avg,3)))\\nax.set_xlabel(\\'GC Content of Non Repeats\\')\\n#ax.set_ylabel(\\'GC Content without Repeats\\')\\nax.set_ylabel(\\'GC Content of Repeats\\')\\nax.legend()\\nax.set_title(\\'GC Content Graph for Bfl\\')\\n'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "#Add in remove repeats\n",
    "fig, ax = plt.subplots(figsize=(8,8))\n",
    "colors = cm.viridis(np.linspace(0,1, numChromosomes))\n",
    "genome = \"\"\n",
    "gcR = []\n",
    "gc = []\n",
    "j=0\n",
    "for index, row in seqobj.iterrows(): \n",
    "        pos_array = []\n",
    "        gc_content_value_array = []\n",
    "        gc_content_value_array_NR = []\n",
    "        gc_content_value_array_R = []\n",
    "    \n",
    "        n_percent_array = []\n",
    "        name = row['id']\n",
    "        seq = row['seq']\n",
    "        for i in range(0,len(seq),step):\n",
    "            subseq = seq[i:i+window]\n",
    "            gc_content = (GC_content_window(subseq))\n",
    "            gc_content_NR = (GC_content_window_NR(subseq))\n",
    "            gc_content_R = (GC_content_window_R(subseq))\n",
    "            n_percent = (null_value_window(subseq))\n",
    "            #start = (i + 1 if (i+1<=len(seq)) else i)\n",
    "            #end = ( i + step if (i+ step<=len(seq)) else len(seq))\n",
    "            if n_percent < nThreshold:\n",
    "                gc_content_value_array.append(gc_content)\n",
    "                gc_content_value_array_NR.append(gc_content_NR)\n",
    "                gc_content_value_array_R.append(gc_content_R)\n",
    "        avg = np.average(gc_content_value_array_R)\n",
    "        ax.scatter(gc_content_value_array_NR, gc_content_value_array_R,color=colors[j], label=str(name), alpha = 0.7)\n",
    "        #ax.plot([0,1],[avg,avg], color=colors[index], linestyle=\"dashed\")\n",
    "        genome = genome+seq\n",
    "        j+=1\n",
    "ax.plot([0,1],[0,1],color='black',linestyle=\"dashed\")\n",
    "avg = GC_content_window_NR(genome)\n",
    "avgR = GC_content_window_R(genome)\n",
    "ax.plot([0,1],[avgR,avgR], color='silver', linestyle=\"dashed\", label=\"Average GC Content of Repeats: \"+str(np.round(avgR,3)))\n",
    "ax.plot([avg,avg],[0,1], color='gold', linestyle=\"dashed\", label=\"Average GC Content of Non Repeats: \"+str(np.round(avg,3)))\n",
    "ax.set_xlabel('GC Content of Non Repeats')\n",
    "#ax.set_ylabel('GC Content without Repeats')\n",
    "ax.set_ylabel('GC Content of Repeats')\n",
    "ax.legend()\n",
    "ax.set_title('GC Content Graph for Bfl')\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6bb3d96c-eb9a-42c3-9352-4e11b979be85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nfor index, row in seqobj.iterrows(): \\n        pos_array = []\\n        gc_content_value_array = []\\n        gc_skew_value_array = []\\n        ag_content_value_array = []\\n        at_skew_value_array = []\\n\\n        pos_array = []\\n        gc_content_value_array_R = []\\n        gc_skew_value_array_R = []\\n        ag_content_value_array_R = []\\n        at_skew_value_array_R = []\\n        \\n        n_percent_array = []\\n        name = row[\\'id\\']\\n        seq = row[\\'seq\\']\\n        for i in range(0,len(seq),step):\\n            subseq = seq[i:i+window]\\n            gc_content = (GC_content_window(subseq))\\n            gc_skew = (GC_skew_window(subseq))\\n            ag_content = (AG_content_window(subseq))\\n            at_skew = (AT_skew_window(subseq))\\n\\n            gc_content_R = (GC_content_window_R(subseq))\\n            gc_skew_R = (GC_skew_window_R(subseq))\\n            ag_content_R = (AG_content_window_R(subseq))\\n            at_skew_R = (AT_skew_window_R(subseq))\\n            \\n            n_percent = (null_value_window(subseq))\\n            #start = (i + 1 if (i+1<=len(seq)) else i)\\n            #end = ( i + step if (i+ step<=len(seq)) else len(seq))\\n            if n_percent < nThreshold:\\n                pos_array.append(i)\\n                gc_content_value_array.append(gc_content)\\n                gc_skew_value_array.append(gc_skew)\\n                ag_content_value_array.append(ag_content)\\n                at_skew_value_array.append(at_skew)\\n\\n                gc_content_value_array_R.append(gc_content_R)\\n                gc_skew_value_array_R.append(gc_skew_R)\\n                ag_content_value_array_R.append(ag_content_R)\\n                at_skew_value_array_R.append(at_skew_R)\\n                \\n                n_percent_array.append(n_percent)\\n        sequence_length = len(seq)\\n        gene_density = np.zeros(sequence_length)\\n        filtered_rows = gtf_dataframe.query(\\'sequence == @name and feature == \"gene\"\\')\\n        for start, end in zip(filtered_rows[\\'start\\'], filtered_rows[\\'end\\']):\\n            gene_density[start:end] += 1\\n        \\n        max_columns = 2**22\\n        if sequence_length > max_columns:\\n            factor = sequence_length // max_columns\\n            new_length = (sequence_length // factor) * factor  # Ensure new length is a multiple of the factor\\n            print(\"Cut off \"+str(sequence_length - ((sequence_length // factor) * factor))+\" bases\")\\n            gene_density_truncated = gene_density[:new_length]  # Truncate the array\\n            gene_density_downsampled = gene_density_truncated.reshape(-1, factor).mean(axis=1)\\n            downsampled = True\\n        else:\\n            gene_density_downsampled = gene_density\\n            downsampled = False\\n        segments = []\\n        colors = []\\n        filteredSC = SC_dataframe.query(\\'sequence == @name\\')\\n        for index, row in filteredSC.iterrows():\\n            start = row[\\'start\\']\\n            end = row[\\'end\\']\\n            segments.append([(start,0), (end,0)])\\n            colors.append(hex_to_rgba(row[\\'color\\']))\\n        lc = LineCollection(segments, colors=colors, linewidths=50)\\n        fig, axs = plt.subplots(7,1,figsize=graphShape, sharex=False)\\n        \\n        axs[0].plot(pos_array, gc_content_value_array, \\'g-\\', label=\\'GC Content\\')\\n        axs[0].plot(pos_array, gc_content_value_array_R, color=\\'DarkOrchid\\', ls = \"--\", lw = 1, label=\\'GC Content of Repeats\\')\\n        axs[0].set_ylabel(\\'GC Content\\')\\n        axs[0].legend()\\n        axs[0].set_xlim(0, sequence_length)\\n    \\n        axs[1].plot(pos_array, gc_skew_value_array, \\'r-\\', label=\\'GC Skew\\')\\n        axs[1].plot(pos_array, gc_skew_value_array_R, color=\\'DarkOrchid\\', ls = \"--\", lw = 1, label=\\'GC Skew of Repeats\\')\\n        axs[1].set_ylabel(\\'GC Skew\\')\\n        axs[1].axhline(0,color=\\'black\\')\\n        axs[1].legend()\\n        axs[1].set_xlim(0, sequence_length)\\n    \\n        axs[2].plot(pos_array, ag_content_value_array, \\'b-\\', label=\\'AG Content\\')\\n        axs[2].plot(pos_array, ag_content_value_array_R, color=\\'DarkOrchid\\', ls = \"--\", lw = 1, label=\\'AG Content of Repeats\\')\\n        axs[2].set_ylabel(\\'AG Content\\')\\n        axs[2].legend()\\n        axs[2].set_xlim(0, sequence_length)\\n    \\n        axs[3].plot(pos_array, at_skew_value_array, \\'y-\\', label=\\'AT Skew\\')\\n        axs[3].plot(pos_array, at_skew_value_array_R, color=\\'DarkOrchid\\', ls = \"--\", lw = 1, label=\\'AT Skew of Repeats\\')\\n        axs[3].set_ylabel(\\'AT Skew\\')\\n        axs[3].axhline(0,color=\\'black\\')\\n        axs[3].legend()\\n        axs[3].set_xlim(0, sequence_length)\\n\\n        axs[4].plot(pos_array, n_percent_array, \\'m-\\', label=\\'Null Percent\\')\\n        axs[4].set_ylabel(\\'Null Percent (%)\\')\\n        axs[4].legend()\\n        axs[4].set_xlabel(\\'Position in Sequence (bp)\\')\\n        axs[4].set_xlim(0, sequence_length)\\n\\n        axs[5].add_collection(lc)\\n        for i in range(0,len(segments)):\\n            axs[5].scatter((segments[i][0][0]+segments[i][1][0])/2,0,color=colors[i])\\n        axs[5].set_ylim(-0.5,0.5)\\n        axs[5].set_xlim(0, sequence_length)\\n        axs[5].set_ylabel(\\'SC Orthologues\\')\\n    \\n        norm=plt.Normalize(gene_density_downsampled.min(), gene_density_downsampled.max())\\n        cmap = plt.get_cmap(\\'viridis\\')\\n        axs[6].imshow(gene_density_downsampled[np.newaxis, :], aspect=\\'auto\\', cmap=cmap, norm=norm)\\n        axs[6].set_yticks([])\\n        axs[6].set_ylabel(\\'Gene Density\\')\\n        \\n        if downsampled:\\n            tick_positions = np.linspace(0, len(gene_density_downsampled), num=10, dtype=int)\\n            tick_labels = np.linspace(0, sequence_length, num=10, dtype=int)\\n            axs[6].set_xticks(tick_positions)\\n            axs[6].set_xticklabels(tick_labels)\\n        else:\\n            axs[6].set_xlim(0, sequence_length)\\n    \\n        #Colorbar\\n        sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\\n        sm.set_array([])\\n        cbar = fig.colorbar(sm, ax=axs[6], orientation=\\'horizontal\\', pad=0.2)\\n        #cbar.set_label(\\'Gene Density\\')\\n    \\n        fig.suptitle(\"Sliding Window Analysis for different metrics on the \"+str(name)+\" scaffold\")\\n        plt.subplots_adjust(hspace=0.5)\\n        #plt.tight_layout()\\n        #plt.subplots_adjust(top=0.95)\\n        plt.show()\\n'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "for index, row in seqobj.iterrows(): \n",
    "        pos_array = []\n",
    "        gc_content_value_array = []\n",
    "        gc_skew_value_array = []\n",
    "        ag_content_value_array = []\n",
    "        at_skew_value_array = []\n",
    "\n",
    "        pos_array = []\n",
    "        gc_content_value_array_R = []\n",
    "        gc_skew_value_array_R = []\n",
    "        ag_content_value_array_R = []\n",
    "        at_skew_value_array_R = []\n",
    "        \n",
    "        n_percent_array = []\n",
    "        name = row['id']\n",
    "        seq = row['seq']\n",
    "        for i in range(0,len(seq),step):\n",
    "            subseq = seq[i:i+window]\n",
    "            gc_content = (GC_content_window(subseq))\n",
    "            gc_skew = (GC_skew_window(subseq))\n",
    "            ag_content = (AG_content_window(subseq))\n",
    "            at_skew = (AT_skew_window(subseq))\n",
    "\n",
    "            gc_content_R = (GC_content_window_R(subseq))\n",
    "            gc_skew_R = (GC_skew_window_R(subseq))\n",
    "            ag_content_R = (AG_content_window_R(subseq))\n",
    "            at_skew_R = (AT_skew_window_R(subseq))\n",
    "            \n",
    "            n_percent = (null_value_window(subseq))\n",
    "            #start = (i + 1 if (i+1<=len(seq)) else i)\n",
    "            #end = ( i + step if (i+ step<=len(seq)) else len(seq))\n",
    "            if n_percent < nThreshold:\n",
    "                pos_array.append(i)\n",
    "                gc_content_value_array.append(gc_content)\n",
    "                gc_skew_value_array.append(gc_skew)\n",
    "                ag_content_value_array.append(ag_content)\n",
    "                at_skew_value_array.append(at_skew)\n",
    "\n",
    "                gc_content_value_array_R.append(gc_content_R)\n",
    "                gc_skew_value_array_R.append(gc_skew_R)\n",
    "                ag_content_value_array_R.append(ag_content_R)\n",
    "                at_skew_value_array_R.append(at_skew_R)\n",
    "                \n",
    "                n_percent_array.append(n_percent)\n",
    "        sequence_length = len(seq)\n",
    "        gene_density = np.zeros(sequence_length)\n",
    "        filtered_rows = gtf_dataframe.query('sequence == @name and feature == \"gene\"')\n",
    "        for start, end in zip(filtered_rows['start'], filtered_rows['end']):\n",
    "            gene_density[start:end] += 1\n",
    "        \n",
    "        max_columns = 2**22\n",
    "        if sequence_length > max_columns:\n",
    "            factor = sequence_length // max_columns\n",
    "            new_length = (sequence_length // factor) * factor  # Ensure new length is a multiple of the factor\n",
    "            print(\"Cut off \"+str(sequence_length - ((sequence_length // factor) * factor))+\" bases\")\n",
    "            gene_density_truncated = gene_density[:new_length]  # Truncate the array\n",
    "            gene_density_downsampled = gene_density_truncated.reshape(-1, factor).mean(axis=1)\n",
    "            downsampled = True\n",
    "        else:\n",
    "            gene_density_downsampled = gene_density\n",
    "            downsampled = False\n",
    "        segments = []\n",
    "        colors = []\n",
    "        filteredSC = SC_dataframe.query('sequence == @name')\n",
    "        for index, row in filteredSC.iterrows():\n",
    "            start = row['start']\n",
    "            end = row['end']\n",
    "            segments.append([(start,0), (end,0)])\n",
    "            colors.append(hex_to_rgba(row['color']))\n",
    "        lc = LineCollection(segments, colors=colors, linewidths=50)\n",
    "        fig, axs = plt.subplots(7,1,figsize=graphShape, sharex=False)\n",
    "        \n",
    "        axs[0].plot(pos_array, gc_content_value_array, 'g-', label='GC Content')\n",
    "        axs[0].plot(pos_array, gc_content_value_array_R, color='DarkOrchid', ls = \"--\", lw = 1, label='GC Content of Repeats')\n",
    "        axs[0].set_ylabel('GC Content')\n",
    "        axs[0].legend()\n",
    "        axs[0].set_xlim(0, sequence_length)\n",
    "    \n",
    "        axs[1].plot(pos_array, gc_skew_value_array, 'r-', label='GC Skew')\n",
    "        axs[1].plot(pos_array, gc_skew_value_array_R, color='DarkOrchid', ls = \"--\", lw = 1, label='GC Skew of Repeats')\n",
    "        axs[1].set_ylabel('GC Skew')\n",
    "        axs[1].axhline(0,color='black')\n",
    "        axs[1].legend()\n",
    "        axs[1].set_xlim(0, sequence_length)\n",
    "    \n",
    "        axs[2].plot(pos_array, ag_content_value_array, 'b-', label='AG Content')\n",
    "        axs[2].plot(pos_array, ag_content_value_array_R, color='DarkOrchid', ls = \"--\", lw = 1, label='AG Content of Repeats')\n",
    "        axs[2].set_ylabel('AG Content')\n",
    "        axs[2].legend()\n",
    "        axs[2].set_xlim(0, sequence_length)\n",
    "    \n",
    "        axs[3].plot(pos_array, at_skew_value_array, 'y-', label='AT Skew')\n",
    "        axs[3].plot(pos_array, at_skew_value_array_R, color='DarkOrchid', ls = \"--\", lw = 1, label='AT Skew of Repeats')\n",
    "        axs[3].set_ylabel('AT Skew')\n",
    "        axs[3].axhline(0,color='black')\n",
    "        axs[3].legend()\n",
    "        axs[3].set_xlim(0, sequence_length)\n",
    "\n",
    "        axs[4].plot(pos_array, n_percent_array, 'm-', label='Null Percent')\n",
    "        axs[4].set_ylabel('Null Percent (%)')\n",
    "        axs[4].legend()\n",
    "        axs[4].set_xlabel('Position in Sequence (bp)')\n",
    "        axs[4].set_xlim(0, sequence_length)\n",
    "\n",
    "        axs[5].add_collection(lc)\n",
    "        for i in range(0,len(segments)):\n",
    "            axs[5].scatter((segments[i][0][0]+segments[i][1][0])/2,0,color=colors[i])\n",
    "        axs[5].set_ylim(-0.5,0.5)\n",
    "        axs[5].set_xlim(0, sequence_length)\n",
    "        axs[5].set_ylabel('SC Orthologues')\n",
    "    \n",
    "        norm=plt.Normalize(gene_density_downsampled.min(), gene_density_downsampled.max())\n",
    "        cmap = plt.get_cmap('viridis')\n",
    "        axs[6].imshow(gene_density_downsampled[np.newaxis, :], aspect='auto', cmap=cmap, norm=norm)\n",
    "        axs[6].set_yticks([])\n",
    "        axs[6].set_ylabel('Gene Density')\n",
    "        \n",
    "        if downsampled:\n",
    "            tick_positions = np.linspace(0, len(gene_density_downsampled), num=10, dtype=int)\n",
    "            tick_labels = np.linspace(0, sequence_length, num=10, dtype=int)\n",
    "            axs[6].set_xticks(tick_positions)\n",
    "            axs[6].set_xticklabels(tick_labels)\n",
    "        else:\n",
    "            axs[6].set_xlim(0, sequence_length)\n",
    "    \n",
    "        #Colorbar\n",
    "        sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "        sm.set_array([])\n",
    "        cbar = fig.colorbar(sm, ax=axs[6], orientation='horizontal', pad=0.2)\n",
    "        #cbar.set_label('Gene Density')\n",
    "    \n",
    "        fig.suptitle(\"Sliding Window Analysis for different metrics on the \"+str(name)+\" scaffold\")\n",
    "        plt.subplots_adjust(hspace=0.5)\n",
    "        #plt.tight_layout()\n",
    "        #plt.subplots_adjust(top=0.95)\n",
    "        plt.show()\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6b65ef79-ab44-4fd1-8073-0ab1d8d6fc8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "numChromosomes = 19\n",
    "filename = 'input_data/genomes/Llo.fna'\n",
    "#filename = 'input_data/genomes/BflAlt.fna'\n",
    "gtffile = 'input_data/gene_rows/Llo.gtf'\n",
    "SCfile = 'input_data/gene_rows_SC/Llo_coordinates.tsv'\n",
    "\n",
    "records = SeqIO.parse(filename, 'fasta')\n",
    "data = [(record.id, str(record.seq), len(record.seq)) for record in records]\n",
    "seqobj = pd.DataFrame(data, columns=['id', 'seq', 'seqlen'])\n",
    "seqobj = seqobj.sort_values('seqlen', ascending=False).head(numChromosomes)\n",
    "\n",
    "columns = [\"sequence\", \"source\", \"feature\", \"start\", \"end\", \"score\", \"strand\", \"frame\", \"attribute\"]\n",
    "gtf_dataframe = pd.read_csv(gtffile, sep = '\\t', comment = '#', header = None, names = columns, dtype={'start': int, 'end': int})\n",
    "\n",
    "SCcolumns = [\"index\", \"Junk\", \"sequence\", \"start\", \"end\", \"color_key\"]\n",
    "SC_dataframe = pd.read_csv(SCfile, sep = '\\t', comment = '#', header = None, names = SCcolumns, dtype={'start': int, 'end': int})\n",
    "SC_dataframe = SC_dataframe[SC_dataframe['color_key'].isin([\"A\",\"B\",\"C\",\"D\",\"E\",\"F\",\"G\",\"H\",\"I\",\"J\",\"K\",\"L\",\"M\",\"N\",\"O\",\"P\",\"Q\",\"R\",\"S\"])]\n",
    "SC_dataframe['color'] = SC_dataframe['color_key'].map(color_mapping)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71541c56-b924-4e54-be4e-aca41a804538",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Just for investigating the bimodal distribution in Llo - centromeres\n",
    "for index, row in seqobj.query('id == \"NC_088313.1\"').iterrows(): \n",
    "        pos_array = []\n",
    "        gc_content_value_array = []\n",
    "        n_percent_array = []\n",
    "        name = row['id']\n",
    "        seq = row['seq']\n",
    "        for i in range(0,len(seq),step):\n",
    "            subseq = seq[i:i+window]\n",
    "            gc_content = (GC_content_window(subseq))\n",
    "            n_percent = (null_value_window(subseq))\n",
    "            if n_percent < nThreshold:\n",
    "                pos_array.append(i)\n",
    "                gc_content_value_array.append(gc_content)\n",
    "                n_percent_array.append(n_percent)\n",
    "        \n",
    "        sequence_length = len(seq)\n",
    "        gene_density = np.zeros(sequence_length)\n",
    "        filtered_rows = gtf_dataframe.query('sequence == @name and feature == \"gene\"')\n",
    "        for start, end in zip(filtered_rows['start'], filtered_rows['end']):\n",
    "            gene_density[start:end] += 1\n",
    "        \n",
    "        max_columns = 2**22\n",
    "        if sequence_length > max_columns:\n",
    "            factor = sequence_length // max_columns\n",
    "            new_length = (sequence_length // factor) * factor  # Ensure new length is a multiple of the factor\n",
    "            print(\"Cut off \"+str(sequence_length - ((sequence_length // factor) * factor))+\" bases\")\n",
    "            gene_density_truncated = gene_density[:new_length]  # Truncate the array\n",
    "            gene_density_downsampled = gene_density_truncated.reshape(-1, factor).mean(axis=1)\n",
    "            downsampled = True\n",
    "        else:\n",
    "            gene_density_downsampled = gene_density\n",
    "            downsampled = False\n",
    "        segments = []\n",
    "        colors = []\n",
    "        filteredSC = SC_dataframe.query('sequence == @name')\n",
    "        for index, row in filteredSC.iterrows():\n",
    "            start = row['start']\n",
    "            end = row['end']\n",
    "            segments.append([(start,0), (end,0)])\n",
    "            colors.append(hex_to_rgba(row['color']))\n",
    "        lc = LineCollection(segments, colors=colors, linewidths=50)\n",
    "        fig, axs = plt.subplots(4,1,figsize=graphShape, sharex=False)\n",
    "        \n",
    "        axs[0].plot(pos_array, gc_content_value_array, 'g-', label='GC Content')\n",
    "        axs[0].set_ylabel('GC Content')\n",
    "        axs[0].legend()\n",
    "        axs[0].set_xlim(0, sequence_length)\n",
    "\n",
    "        for start, end in zip(filtered_rows['start'],filtered_rows['end']):\n",
    "            gcctent = GC_content_window(seq[start:end])\n",
    "            if gcctent > 0.5 and gcctent < 0.6:\n",
    "                axs[1].scatter((start+end)/2,0)\n",
    "        axs[1].set_ylabel('Between 0.5 to 0.6')\n",
    "        axs[1].set_xlim(0, sequence_length)\n",
    "\n",
    "        axs[2].add_collection(lc)\n",
    "        for i in range(0,len(segments)):\n",
    "            axs[2].scatter((segments[i][0][0]+segments[i][1][0])/2,0,color=colors[i])\n",
    "        axs[2].set_ylim(-0.5,0.5)\n",
    "        axs[2].set_xlim(0, sequence_length)\n",
    "        axs[2].set_ylabel('SC Orthologues')\n",
    "    \n",
    "        norm=plt.Normalize(gene_density_downsampled.min(), gene_density_downsampled.max())\n",
    "        cmap = plt.get_cmap('viridis')\n",
    "        axs[3].imshow(gene_density_downsampled[np.newaxis, :], aspect='auto', cmap=cmap, norm=norm)\n",
    "        axs[3].set_yticks([])\n",
    "        axs[3].set_ylabel('Gene Density')\n",
    "        \n",
    "        if downsampled:\n",
    "            tick_positions = np.linspace(0, len(gene_density_downsampled), num=10, dtype=int)\n",
    "            tick_labels = np.linspace(0, sequence_length, num=10, dtype=int)\n",
    "            axs[3].set_xticks(tick_positions)\n",
    "            axs[3].set_xticklabels(tick_labels)\n",
    "        else:\n",
    "            axs[3].set_xlim(0, sequence_length)\n",
    "    \n",
    "        #Colorbar\n",
    "        sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "        sm.set_array([])\n",
    "        cbar = fig.colorbar(sm, ax=axs[3], orientation='horizontal', pad=0.2)\n",
    "        #cbar.set_label('Gene Density')\n",
    "    \n",
    "        fig.suptitle(\"Sliding Window Analysis for different metrics on the \"+str(name)+\" scaffold\")\n",
    "        plt.subplots_adjust(hspace=0.5)\n",
    "        #plt.tight_layout()\n",
    "        #plt.subplots_adjust(top=0.95)\n",
    "        plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
